### ISSUE TITLE
Questions from a sensor-fusion newbie

### ISSUE THREAD

**Issue by TheButlah** [daniel-om-weber/riann#2]
Hi,

I'm looking to use your nn in a project for estimating 6dof AHRS. I have a couple of questions I was wondering if you could answer:

From #1:
> The optimization function of the neural network ignores the heading of the target orientation which results in a not defined heading.

Does this mean that this SFA does not predict yaw? To be clear, I don't care about yaw with respect to earth's magnetic field, just total change in yaw with respect to when the sensor finished its initial calibration.

Also, the paper discussed the use of the SFA in an Nvidia Jetson. I've used a Jetson before and they are quite beefy compared to an embedded microcontroller like a ESP8266 for example. Do you have any opinions on if this SFA could run on that sort of platform (we are using the Wemos D1 mini for reference)? I was thinking of opting for 40 neurons per layer since that seems like the point of diminishing returns in the paper

I don't see the code used to actually train the model anywhere, just the .onnx file. Is there an associated model implementation along with training code?

 Does the code support the time aware network? In the paper it is described that `dt` which is the time since the last sample is provided as an input, but the python class here accepts `fs` which is the sample rate of the data. I will have a varying `dt`, so how would I convert my varying `dt`s to `fs`? Is it just `fs = 1/dt`?

**Comment by daniel-om-weber** [comment]
Hi,
I am glad to answer your questions.

> From #1:
> 
> > The optimization function of the neural network ignores the heading of the target orientation which results in a not defined heading.
> 
> Does this mean that this SFA does not predict yaw? To be clear, I don't care about yaw with respect to earth's magnetic field, just total change in yaw with respect to when the sensor finished its initial calibration.

This is correct. Yaw is the object's angle relative to its coordinate system, and the heading is the angle relative to a global coordinate system. If you are interested in the change of yaw, you may use the yaw axis of the gyroscope. It measures the yaw rate directly.
 
> Also, the paper discussed the use of the SFA in an Nvidia Jetson. I've used a Jetson before and they are quite beefy compared to an embedded microcontroller like a ESP8266 for example. Do you have any opinions on if this SFA could run on that sort of platform (we are using the Wemos D1 mini for reference)? I was thinking of opting for 40 neurons per layer since that seems like the point of diminishing returns in the paper
> 
> I don't see the code used to actually train the model anywhere, just the .onnx file. Is there an associated model implementation along with training code?

RIANN is a ready-to-use attitude estimator which is not meant to be trained on new data. The training relies on code, which we aim to release in the future. An ESP8266 is a challenging target for a neural network-based estimator. The main limiting factor is the 80 kB of Ram. Further optimizations towards tiny models may change this, but the size analysis of the paper shows that conventional filters are more accurate and faster on those platforms.

> Does the code support the time aware network? In the paper it is described that `dt` which is the time since the last sample is provided as an input, but the python class here accepts `fs` which is the sample rate of the data. I will have a varying `dt`, so how would I convert my varying `dt's to `fs`? Is it just `fs = 1/dt`?

The code supports time-aware execution with a sampling rate fixed over a sequence. Technically you can input `fs = 1/dt` but I can not predict whether this will perform well, as the model has been trained with fixed sampling rates.

**Comment by TheButlah** [comment]
Is it generally true that 6dof SFAs do not help with drift in yaw at all? I had thought that they would be able to help at least a little

**Comment by daniel-om-weber** [comment]
6d SFAs may reduce but not eliminate the drift by estimating and compensating for systematic errors. They cannot deliver dependable results after long periods because the problem is not observable.
